File: Algorithms/8. Graph/1. Graph Traversal/Applications/connectivity check/connectCheck.cpp
// this program demonstrate using dfs to check the connectivity of graph
# include <iostream>
# include <vector>
using namespace std;

// Graph class
class Graph {
  vector< vector<int> > adjacencyList; // adjacency list implementation
  vector<bool> visited; // track visited node
public:
  // constructor to resize the adjacencyList and Visited vector
  Graph(int size) {
    adjacencyList.resize(size);
    visited.resize(size);
  }

  // add adjacencyList to the graph with given current node
  void addEdge(int node1, int node2) {
    adjacencyList[node1].push_back(node2);
  }

  // Depth first search algorithm
  void DFS(int start) { 
    if (visited[start]) return; // if visted, return
    // cout << start << " "; // for debug purpose
    visited[start] = true;
    // process node start's connected node in the list
    for (auto u: adjacencyList[start])
      DFS(u); // recurse on the connected node until all visited
  }

  // check connectivity
  bool connectCheck(int start) {
    // using dfs to traverse the graph and check connectivity
    DFS(start);
    for (bool b : visited)
      if (!b)
        return false;
    return true;
  }
};

// output for connectCheck function
void connectStatus(bool b) {
  if (!b) cout << "Graph not fully connected!" << endl;
  else cout << "Graph is fully connected!" << endl;
}

int main() {
  // create a graph called g and add adjacencyList to it
  Graph graph1(4);
  graph1.addEdge(0,1);
  graph1.addEdge(0,2);
  graph1.addEdge(1,0);
  graph1.addEdge(2,0);
  graph1.addEdge(2,3);
  graph1.addEdge(3,2);

  // check connectivity
  cout << "- Testing connectivity for graph1:\n";
  connectStatus(graph1.connectCheck(0));
  cout << endl;

    // create a graph called g and add adjacencyList to it
  Graph graph2(5);
  graph2.addEdge(0,1);
  graph2.addEdge(0,2);
  graph2.addEdge(2,0);
  graph2.addEdge(2,1);
  graph2.addEdge(3,4);
  graph2.addEdge(4,3);

  // check connectivity
  cout << "- Testing connectivity for graph2:\n";
  connectStatus(graph2.connectCheck(0));
}

File: Algorithms Techniques/8. Graph/1. Graph Traversal/Applications/connectivity check/connectCheck.py
# python implementation of DFS traversal of a graph
class Graph:
  # constructor
  def __init__(self, size):
    # use list comprehension to create empty adjacency list and 
    # visited boolean tracker
    self.visited = [False for i in xrange(size)]
    self.adj_list = [ [] for i in xrange(size)]
  
  # add edges to the graph
  def add_edge(self, node1, node2):
    self.adj_list[node1].append(node2)

  # dfs algorithm
  def dfs(self, start):
    # if visited, go back one level
    if self.visited[start]: 
      return 
    # print(str(start) + " "), # use it for debug
    self.visited[start] = True
    for node in self.adj_list[start]:
      self.dfs(node) # recursively visit connected node until return
  
  # check the connectivity
  def connect_check(self, start):
    self.dfs(start)
    for b in self.visited:
      if not b:
        return False
    return True

# output the grahp connectivity status
def connect_status(b):
  if not b: print("Graph not fully connected!")
  else: print("Graph is fully connected!")

# =================================
# testing the connectivity of graph
# =================================

# create graph with 4 nodes
graph1 = Graph(4)

# add edges with nodes to the graph
graph1.add_edge(0,1)
graph1.add_edge(0,2)
graph1.add_edge(1,0)
graph1.add_edge(2,0)
graph1.add_edge(2,3)
graph1.add_edge(3,2)

# search the graph using dfs algorithm
print("- Testing connectivity for graph1:")
connect_status(graph1.connect_check(0))
print

# create graph with 5 nodes
graph2 = Graph(5)

# add edges with nodes to the graph
graph2.add_edge(0,1)
graph2.add_edge(0,2)
graph2.add_edge(2,0)
graph2.add_edge(2,1)
graph2.add_edge(3,4)
graph2.add_edge(4,3)

# search the graph using dfs algorithm
print("- Testing connectivity for graph2:")
connect_status(graph2.connect_check(0))